sw/device/silicon_creator/lib/sigverify_dynamic_functest.c

```diff
@@ -0,0 +1,65 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/lib/base/memory.h"
+#include "sw/device/lib/testing/check.h"
+#include "sw/device/silicon_creator/lib/base/sec_mmio.h"
+#include "sw/device/silicon_creator/lib/drivers/hmac.h"
+#include "sw/device/silicon_creator/lib/sigverify.h"
+#include "sw/device/silicon_creator/lib/sigverify_tests/sigverify_testvectors.h"
+#include "sw/device/silicon_creator/lib/test_main.h"
+
+// Index of the test vector currently under test
+static uint32_t test_index;
+
+rom_error_t sigverify_test(void) {
+  sigverify_test_vector_t testvec = sigverify_tests[test_index];
+
+  // Extract the digest from the encoded message
+  hmac_digest_t digest;
+  memcpy(digest.digest, testvec.encoded_msg,
+         kHmacDigestNumWords * sizeof(uint32_t));
+
+  uint32_t flash_exec = 0;
+  rom_error_t result = sigverify_rsa_verify(&testvec.sig, &testvec.key, &digest,
+                                            kLcStateRma, &flash_exec);
+
+  rom_error_t test_result;
+  if (testvec.valid) {
+    CHECK(flash_exec == kSigverifyFlashExec);
+    if (result == kErrorOk) {
+      test_result = kErrorOk;
+    } else {
+      // Signature verification failed when it was expected to pass.
+      LOG_ERROR("Error while attempting to verify a valid signature.");
+      LOG_INFO("Test notes: %s", testvec.comment);
+      test_result = result;
+    }
+  } else {
+    CHECK(flash_exec == UINT32_MAX);
+    if (result == kErrorOk) {
+      // Signature verification passed when it was expected to fail.
+      LOG_ERROR("Invalid signature passed verification.");
+      LOG_INFO("Test notes: %s", testvec.comment);
+      test_result = kErrorUnknown;
+    } else {
+      test_result = kErrorOk;
+    }
+  }
+
+  return test_result;
```

Comment:
This part could be slightly simplified but I'm OK if you want to keep `test_result` and assignments in different branches.
```suggestion
  if (testvec.valid) {
    CHECK(flash_exec == kSigverifyFlashExec);
    if (result != kErrorOk) {
      // Signature verification failed when it was expected to pass.
      LOG_ERROR("Error while attempting to verify a valid signature.");
      LOG_INFO("Test notes: %s", testvec.comment);
    }
  } else {
    CHECK(flash_exec == UINT32_MAX);
    if (result == kErrorOk) {
      // Signature verification passed when it was expected to fail.
      LOG_ERROR("Invalid signature passed verification.");
      LOG_INFO("Test notes: %s", testvec.comment);
      return kErrorUnknown;
    }
  }
  return result;
```

Reply:
It's probably a matter of personal style, but I was finding the return paths a little too easy to confuse and though that the separate `test_result` variable was clearer (although more verbose). For instance, the suggestion above will return a non-OK result when the invalid signature is (correctly) failed, so there would need to be another return statement for the case when `testvec.valid = false` and `result != kErrorOk`, and the very last `return result` would only be reachable if the test was valid.